File: maps/src/main/java/it/localhostsoftware/maps/model/LatLngBounds.java
package it.localhostsoftware.maps.model;

import android.content.Context;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import it.localhostsoftware.maps.google.model.GoogleLatLngBounds;

public abstract class LatLngBounds<LB> {
    private final LB lb;

    public LatLngBounds(LB lb) {
        this.lb = lb;
    }

    public LB getLatLngBounds() {
        return lb;
    }

    public abstract LatLng<?> getSouthwest();

    public abstract LatLng<?> getNortheast();

    public abstract boolean contains(LatLng<?> var1);

    public abstract LatLngBounds<?> including(LatLng<?> var1);

    public abstract LatLng<?> getCenter();

    public abstract static class Builder<B> {
        public static Builder<?> getInstance(Context context) {
            if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS)
                return new GoogleLatLngBounds.GoogleBuilder(new com.google.android.gms.maps.model.LatLngBounds.Builder());
            else return null;
        }

        private final B b;

        public Builder(B b) {
            this.b = b;
        }

        public B getBuilder() {
            return b;
        }

        public abstract LatLngBounds.Builder<?> include(LatLng<?> var1);

        public abstract LatLngBounds<?> build();
    }
}

File: maps/src/main/java/it/localhostsoftware/maps/model/LatLngBounds.kt
package it.localhostsoftware.maps.model

import android.content.Context
import it.localhostsoftware.maps.MobileServices
import it.localhostsoftware.maps.getMobileServices
import it.localhostsoftware.maps.google.model.GoogleLatLngBounds.GoogleBuilder
import it.localhostsoftware.maps.huawei.model.HuaweiLatLngBounds.HuaweiBuilder

abstract class LatLngBounds<LB, LL : LatLng<*>>(val lb: LB) {
    abstract val southwest: LL
    abstract val northeast: LL
    abstract val center: LL
    abstract operator fun contains(var1: LL): Boolean
    abstract fun including(var1: LL): LatLngBounds<*, *>
    abstract class Builder<B, LL : LatLng<*>>(val builder: B) {
        companion object {
            fun getInstance(c: Context): Builder<*, *> =
                when (c.getMobileServices()) {
                    MobileServices.GOOGLE -> GoogleBuilder(com.google.android.gms.maps.model.LatLngBounds.Builder())
                    MobileServices.HUAWEI -> HuaweiBuilder(com.huawei.hms.maps.model.LatLngBounds.Builder())
                    else -> throw IllegalStateException()
                }
        }

        abstract fun include(var1: LL): Builder<*, *>
        abstract fun build(): LatLngBounds<*, *>
    }
}